mesonbuild/coredata.py

```diff
@@ -1215,6 +1211,7 @@ def add_to_argparse(self, name: str, parser: argparse.ArgumentParser, help_suffi
     (OptionKey('werror'),          BuiltinOption(UserBooleanOption, 'Treat warnings as errors', False, yielding=False)),
     (OptionKey('wrap_mode'),       BuiltinOption(UserComboOption, 'Wrap mode', 'default', choices=['default', 'nofallback', 'nodownload', 'forcefallback', 'nopromote'])),
     (OptionKey('force_fallback_for'), BuiltinOption(UserArrayOption, 'Force fallback for those subprojects', [])),
+    (OptionKey('backend_startup_project'), BuiltinOption(UserStringOption, 'The default Visual Studio active project', '')),
```

Comment:
this won't actually do what you want (it's a bit surprising I know), you need `OptionKey.from_string('backend_startup_project')`, or `OptionKey('startup_project', _type=OptionType.BACKEND)`

Reply:
Oh! How so? It seems to do what I was after:

```
project('startup_test', ['cpp'], default_options : ['backend_startup_project=a'])
executable('a', 'main.cpp')
executable('b', 'main.cpp')
executable('c', 'main.cpp')
executable('d', 'main.cpp')
```
![a](https://user-images.githubusercontent.com/9592489/142941985-16225e50-e08d-4a91-9821-f94f6f6fbb8b.jpg)

Change to `d`:

![d](https://user-images.githubusercontent.com/9592489/142942033-032aa633-6dd0-431e-beba-0018c9a4bafe.jpg)

Now what is weird but unrelated to the change (does it with master) is `b` or `c` lose `d`!

![b](https://user-images.githubusercontent.com/9592489/142942238-e38d986a-17c3-46c2-8819-e5324287bc7b.jpg)

I might investigate that tomorrow. I've tripped over it before and just reordered things to make it work, but can't be intentional...

Reply:
OK I stepped through it, looks like 

https://github.com/mesonbuild/meson/blob/5163a02fec8836a55952353fb598ac6ced3b6a3c/mesonbuild/mesonlib/universal.py#L2071-L2072

https://github.com/mesonbuild/meson/blob/5163a02fec8836a55952353fb598ac6ced3b6a3c/mesonbuild/mesonlib/universal.py#L2028-L2030

is setting the type. If being explicit is still preferred I can make that change!

Cheers

Reply:
The problem is the key should be stored as `OptionKey('startup_project', _type=OptionType.BACKEND)`, which `from_string` handles correctly, but if you just pass the string straight to the initializer it gets stored as `OptionKey('backup_startup_project', _type=OptionType.BACKUP)`.

This is one of those cases where I wish that python had private constructors, using the initializer directly is fragile.

Reply:
Well I'm more of a Typescript programmer than a Python programmer, and I'm not much of a Typescript programmer (as I may have mentioned)...

```
class OptionKey:
...
    def __init__(self, name: str, subproject: str = '',
                 machine: MachineChoice = MachineChoice.HOST,
                 lang: T.Optional[str] = None,
                 module: T.Optional[str] = None,
                 _type: T.Optional[OptionType] = None):
        # the _type option to the constructor is kinda private. We want to be
        # able tos ave the state and avoid the lookup function when
        # pickling/unpickling, but we need to be able to calculate it when
        # constructing a new OptionKeyhkj
        object.__setattr__(self, 'name', name)
        object.__setattr__(self, 'subproject', subproject)
        object.__setattr__(self, 'machine', machine)
        object.__setattr__(self, 'lang', lang)
        object.__setattr__(self, 'module', module)
        object.__setattr__(self, '_hash', hash((name, subproject, machine, lang, module)))
        if _type is None:
            _type = _classify_argument(self)
        object.__setattr__(self, 'type', _type)
```

Stepping through with `(OptionKey('backend_startup_project'), BuiltinOption(UserStringOption, 'The default Visual Studio active project', ''))` the initializer is called with:

```
name == 'backend_startup_project'
subproject == ''
machine == MachineChoice.HOST
lang == None
module == None
_type == None
```
Then `_classify_argument` fixes up type to `OptionType.BACKEND`.

With `(OptionKey.from_string('backend_startup_project'), BuiltinOption(UserStringOption, 'The default Visual Studio active project', ''))` it ends up called the same:

```
name == 'backend_startup_project'
subproject == ''
machine == MachineChoice.HOST
lang == None
module == None
_type == None
```

When you said "or `OptionKey('startup_project', _type=OptionType.BACKEND)`" did you mean `backend_startup_project`? It doesn't seem to prefix with `backend_` if `_type` is `BACKEND` so doesn't solve the problem (option is still unrecognised - in fact worse as not recognised with `backend=vs` due to being removed from `init_backend_options`!).

`OptionKey('backend_startup_project', _type=OptionType.BACKEND)` works, and saves the call to `_classify_argument` but other than that I can't see the difference :(

Sorry if I'm being dense.

Reply:
You have found a bug! `from_string` *should* be splitting `backend` off of the string in the same way it would split `b_` from a builtin option, which it doesn't either. It only seems to be properly splitting language... Sigh

Reply:
I guess then for the moment it doesn't matter which way you do it, they'll both work the same. I'll have to sort out the bug another time.

Reply:
Ah I see, makes sense now!

In that case, is it actually best to use `OptionKey.from_string` as it'll work now and still be correct when fixed? And obviously change https://github.com/mesonbuild/meson/blob/bd9d9818711eaba18b774e17c9ce405c5020c6f9/mesonbuild/backend/vs2010backend.py#L449 too.

Also, it struck me this is actually a regression in 0.60. Previously it was only `rewrite` that complained (#8174) but as of 0.60 it's `setup` (so explains recent #9442). Not sure if that warrants a fix getting out sooner rather than later and whether the doc snippet is correct?